cudamapper/src/main.cu

```diff
@@ -389,403 +182,331 @@ void align_overlaps(DefaultDeviceAllocator allocator,
-        workers.push_back(std::thread(
-            [&, device_id]() {
-                cudaSetDevice(device_id);
-                while (ranges_idx < get_size<int>(query_target_ranges))
-                {
-                    int range_idx = ranges_idx.fetch_add(1);
-                    //Need to perform this check again for thread-safety
-                    if (range_idx < get_size<int>(query_target_ranges))
-                    {
-                        //compute overlaps takes a range of read_ids and a device ID and uses
-                        //that device to compute the overlaps. It prints overlaps to stdout.
-                        //since multiple worker threads are running stdout is guarded
-                        //by a mutex (`std::mutex overlaps_writer_mtx`)
-                        compute_overlaps(query_target_ranges[range_idx],
-                                         device_id,
-                                         cuda_streams[device_id]);
-                    }
-                }
-            }));
-    }
-
-    // Wait for all per-device threads to terminate
-    for (auto& worker_thread : workers)
-    {
-        worker_thread.join();
-    }
-
-    // Wait for all futures (for overlap writing) to return
-    while (num_overlap_chunks_to_print != 0)
-    {
-        std::this_thread::sleep_for(std::chrono::milliseconds(200));
+        worker_threads.emplace_back(worker_thread_function,
+                                    device_id,
+                                    std::ref(batches_of_indices),
+                                    std::ref(parameters),
+                                    std::ref(output_mutex),
+                                    cuda_streams[device_id],
+                                    number_of_total_batches,
+                                    std::ref(number_of_processed_batches));
     }
 
-    // After last writer_thread_function has decreased num_overlap_chunks_to_print it will still take
-    // some time to destroy its pointer to indices
-    // TODO: this is a workaround, this part of code will be significantly changed with new index caching
-    std::this_thread::sleep_for(std::chrono::milliseconds(200));
-
-    device_index_cache.clear();
-
-    // streams can only be destroyed once all writer threads have finished as they hold references
-    // to indices which have device arrays associated with streams
-    for (cudaStream_t cuda_stream : cuda_streams)
+    // wait for all work to be done
+    for (int32_t device_id = 0; device_id < parameters.num_devices; ++device_id)
     {
-        CGA_CU_CHECK_ERR(cudaStreamSynchronize(cuda_stream));
-        CGA_CU_CHECK_ERR(cudaStreamDestroy(cuda_stream));
+        CGA_CU_CHECK_ERR(cudaSetDevice(device_id));
+        worker_threads[device_id].join();
+        CGA_CU_CHECK_ERR(cudaStreamDestroy(cuda_streams[device_id])); // no need to sync, it should be done at the end of worker_threads
```

Comment:
Should we put a device sync here anyway?

Reply:
I'm not quite sure I'm following @mimaric.
I think it is a good idea that a worker thread only finishes when all the work that was assigned to it is done. So this implies a `cudaStreamSynchronize()` for each thread on its assigned stream. So, I don't see a reason for a device sync here.

Reply:
Yes, that's exactly what I meant.
@vellamike Are you ok with leaving it as it is?

Reply:
Yes